core/common/src/main/java/alluxio/underfs/SeekableUnderFileInputStream.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.underfs;
+
+import alluxio.Seekable;
+
+import com.google.common.base.Preconditions;
+
+import java.io.IOException;
+import java.io.InputStream;
+
+/**
+ * A seekable under file input stream wrapper that encapsulates the under file input stream.
+ * Subclasses of this abstract class should implement the
+ * {@link SeekableUnderFileInputStream#seek(long)} and the reposition the wrapped input stream.
+ *
+ * The {@link #mResourceId} is used by Alluxio worker to cache for reuse.
+ */
+public abstract class SeekableUnderFileInputStream extends InputStream implements Seekable {
```

Comment:
can we extends [`FilteredInputStream`](https://docs.oracle.com/javase/7/docs/api/java/io/FilterInputStream.html) from JDK to reduce the code duplication?

Reply:
Cool. Good to know.